tests/common/plugins/conditional_mark/__init__.py

```diff
@@ -0,0 +1,295 @@
+    except Exception as e:
+        logger.error('Failed to evaluate condition, raw_condition={}, condition_str={}'.format(
+            condition,
+            condition_str))
+        return False
+
+
+def evaluate_conditions(conditions, basic_facts):
+    """Evaluate all the condition strings.
+
+    Evaluate a single condition or multiple conditions. If multiple conditions are supplied, apply AND logical operation
+    to all of them.
+
+    Args:
+        conditions (str or list): Condition string or list of condition strings.
+        basic_facts (dict): A one level dict with basic facts. Keys of the dict can be used as variables in the
+            condition string evaluation.
+
+    Returns:
+        bool: True or False based on condition strings evaluation result.
+    """
+    if isinstance(conditions, list):
+        # Apply 'AND' operation to list of conditions
+        # Personally, I think it makes more sense to apply 'AND' logical operation to a list of conditions.
+        return all([evaluate_condition(c, basic_facts) for c in conditions])
+    else:
+        if conditions is None or conditions.strip() == '':
+            return True
+        return evaluate_condition(conditions, basic_facts)
+
+
+def pytest_collection(session):
+    """Hook for loading conditions and basic facts.
+
+    The pytest session.config.cache is used for caching loaded conditions and basic facts. The cached content can
+    persis across different pytest sessions. You can add pytest command line option '--clear-cache' to always clear
+    cache before pytest run.
+
+    Args:
+        session (obj): Pytest session object.
+    """
+    conditions = load_conditions(session)
+    if conditions:
+        session.config.cache.set('TESTS_MARK_CONDITIONS', conditions)
+
+        # Only load basic facts if conditions are defined.
+        basic_facts = load_basic_facts(session)
+        session.config.cache.set('BASIC_FACTS', basic_facts)
+
+
+def pytest_collection_modifyitems(session, config, items):
+    """Hook for adding marks to test cases based on conditions defind in a centralized file.
+
+    Args:
+        session (obj): Pytest session object.
+        config (obj): Pytest config object.
+        items (obj): List of pytest Item objects.
+    """
+
+    conditions = config.cache.get('TESTS_MARK_CONDITIONS', None)
+    if not conditions:
+        logger.debug('No skip condition is defined')
+        return
+    logger.debug('Predefined skip conditions\n{}'.format(json.dumps(conditions, indent=2)))
+
+    basic_facts = config.cache.get('BASIC_FACTS', None)
+    if not basic_facts:
+        logger.debug('Not able to get basic facts')
+        return
+    logger.info('Available basic facts that can be used in conditional skip:\n{}'.format(
+        json.dumps(basic_facts, indent=2)))
+
+    for item in items:
+        logger.info('Processing: {}'.format(item.nodeid))
+        longest_match = find_longest_match(item.nodeid, conditions.keys())
```

Comment:
Will the LPM be a problem if we want to skip the whole test file and xfail for a specific test case?
```yaml
feature_a/test_file_1.py:
  skip:
    reason: "all testcases in test_file_1.py should be skipped for 202012 image"
    conditions:
      - "release in ['202012']"
feature_a/test_file_1.py::testcase_i:
  xfail:
    reason: "testcase_i are suppose to fail because an issue"
    conditions:
      - https://github.com/Azure/sonic-mgmt/issues/1234
```

Reply:
No. Marks are only added to test cases, not added to whole test file.

Assume we have below test script:

feature_a/test_file_1.py:
```
def testcase_1

def testcase_2

def testcase_i
```
In this example, `testcase_1` and `testcase_2` will have nodeid like `feature_a/test_file_1.py::testcase_1` and `feature_a/test_file_1.py::testcase_2`. They will match entry `feature_a/test_file_1.py`. So, the `skip` mark will be added to `testcase_1` and `testcase_2` when `release in ['201911']`.
For `testcase_i`, its nodeid will be `feature_a/test_file_1.py::testcase_i`. Then it will only match `feature_a/test_file_1.py::testcase_i`. The `xfail` mark will be added to `testcase_i` when the Github issue is still open. Entry `feature_a/test_file_1.py` also matches its nodeid. But, because it is not the longest match, it will simply be ignored.

In a summary, under such scenario, the `skip` mark will be conditionally added to `testcase_1` and `testcase_2`. The `xfail` mark will be conditionally added to `testcase_i`.

Reply:
OK, got it, so the mark condition statement is applied only to the test case level.

Reply:
Could you please give this an emphasis in the `README`?

Reply:
> OK, got it, so the mark condition statement is applied only to the test case level. If the test case is parameterized, the condition file should be like:
> 
> ```yaml
> feature_a/test_file_1.py::testcase_i[1-1-1]:
>   skip:
>     reason: "testcase_i should be skipped for 201911 image"
>     conditions:
>       - "release in ['201911']"
> feature_a/test_file_1.py::testcase_i[2-2-2]:
>   skip:
>     reason: "testcase_i should be skipped for 201911 image"
>     conditions:
>       - "release in ['201911']"
>   xfail:
>     reason: "testcase_i[2-2-2] are suppose to fail because an issue"
>     conditions:
>       - https://github.com/Azure/sonic-mgmt/issues/1234
> ```

Yes, we can add different marks for different params of same test case.